libc-bottom-half/sources/send.c

```diff
@@ -0,0 +1,249 @@
+#include <sys/socket.h>
+
+#include <errno.h>
+
+#include <wasi/api.h>
+#include <wasi/descriptor_table.h>
+#include <wasi/sockets_utils.h>
+
+static ssize_t tcp_sendto(tcp_socket_t *socket, const uint8_t *buffer,
+			  size_t length, int flags, const struct sockaddr *addr,
+			  socklen_t addrlen)
+{
+	const int supported_flags = MSG_DONTWAIT | MSG_NOSIGNAL;
+	if ((flags & supported_flags) != flags) {
+		errno = EOPNOTSUPP;
+		return -1;
+	}
+
+	if (addr != NULL || addrlen != 0) {
+		errno = EISCONN;
+		return -1;
+	}
+
+	tcp_socket_state_connected_t connection;
+	if (socket->state.tag == TCP_SOCKET_STATE_CONNECTED) {
+		connection = socket->state.connected;
+	} else {
+		errno = ENOTCONN;
+		return -1;
+	}
+
+	bool should_block = socket->blocking;
+	if ((flags & MSG_DONTWAIT) != 0) {
+		should_block = false;
+	}
+
+	if ((flags & MSG_NOSIGNAL) != 0) {
+		// Ignore it. WASI has no Unix-style signals. So effectively,
+		// MSG_NOSIGNAL is always the case, whether it was explicitly
+		// requested or not.
+	}
+
+	streams_borrow_output_stream_t tx_borrow =
+		streams_borrow_output_stream(connection.output);
+	while (true) {
+		streams_stream_error_t error;
+		uint64_t count;
+		if (!streams_method_output_stream_check_write(tx_borrow, &count,
+							      &error)) {
+			// TODO wasi-sockets: wasi-sockets has no way to recover stream errors yet.
+			errno = EPIPE;
+			return -1;
+		}
+
+		if (count) {
+			count = count < length ? count : length;
+			preview2_list_u8_t list = { .ptr = (uint8_t *)buffer,
+						    .len = count };
+			if (!streams_method_output_stream_write(
+				    tx_borrow, &list, &error)) {
+				// TODO wasi-sockets: wasi-sockets has no way to recover TCP stream errors yet.
+				errno = EPIPE;
+				return -1;
+			} else {
+				return count;
+			}
```

Comment:
> implement POSIX send semantics where, when send returns successfully, it has actually transferred the number of bytes it has claimed to have transferred.

That's not quite right. When `send` returns successfully, it doesn't mean anything. I tried to capture this in https://github.com/WebAssembly/wasi-io/issues/73
> the OS' non-blocking send implementation doesn't perform any IO itself. Most likely all it does is move the data from user space to a kernel queue for it to be put on the wire at a later moment. If send returns an error, it's likely that it has nothing to do with the current send call, but instead it is a delayed notification that one of the previous sends failed. Or put differently; if send returns success, it doesn't mean the data was successfully sent. It only means ownership of the buffer has been transferred from the application to the kernel.

IMO, filesystem fds and internet socket fds are different beasts entirely. Sockets are already stream-like by nature. The fact that `write` on filesystem descriptors need flushing after every write is because WASI exposes the file's contents as a buffered stream. Yet, POSIX promises consistency between other readers/writers on that same file. Hence the flush.

I don't have many examples nor counterexamples, but I'd say that within the general concept of "streams", the filesystem's behavior of "flush every write" is more an exception than the norm.

Reply:
My concern here is not about "is the write visible to anyone" or "has the peer received the bytes". This is just about "would POSIX expect us to report an error from this `send` call?". Calling `flush` should ensure that if the implementation has an `ECONNRESET` waiting for us that a synchronous `send` would have observed, that we observe it too.

One could argue that if we don't observe it in the current `send` call, we'll just report it in the next `send` call, and that that's acceptable. And maybe it is; my thought was just to err on the side of caution for now.

Reply:
> This is just about "would POSIX expect us to report an error from this send call?".

How would this work on non-blocking sockets?